Lib/typing.py

```diff
@@ -823,6 +874,117 @@ def __init__(self, name, *constraints, bound=None,
+      C[int, str]  # Also fine
+      C[()]        # Even this is fine
+
+    For more details, see PEP 646.
+    """
+
+    def __init__(self, name):
+        self._name = name
+        self._unpacked = UnpackedTypeVarTuple(name, self)
+
+    def __iter__(self):
+        yield self._unpacked
+
+    def __repr__(self):
+        return self._name
+
+
+class UnpackedTypeVarTuple(_Final, _Immutable, _root=True):
+
+    def __init__(self, name, packed):
+        self._name = name
+        self._packed = packed
+        # __parameters__ will be use to figure out what the __parameters__
+        # of e.g. `tuple[*Ts]` should be. In the case of an unpacked
+        # TypeVarTuple, it's unclear whether this should refer to the
+        # UnpackedTypeVarTuple or the TypeVarTuple itself. We somewhat
+        # arbitrarily decide to go for the latter.
+        self.__parameters__ = (packed,)
+
+    def __repr__(self):
+        return '*' + self._name
+
+
+class StarredTuple(_Final, _Immutable, _root=True):
+    """Implementation of starred tuple for older versions of Python.
+
+    A starred tuple is e.g. tuple[*tuple[int, str]]. In newer versions of
+    Python, this is handled through a change to `tuple` itself.
+    """
+
+    def __init__(self, tup):
+        self.__origin__ = tuple
+        self._tuple = tup
+        self.__args__ = tup.__args__
+        self.__parameters__ = tup.__parameters__
+
+    def __repr__(self):
+        return '*' + repr(self._tuple)
+
+    def __eq__(self, other):
+        if isinstance(other, StarredTuple):
+            return self._tuple == other._tuple
+        else:
+            return False
+
+
+class Unpack(_Final, _Immutable, _root=True):
+    """Implementation of the unpack operator for older versions of Python.
+
+    The type unpack operator takes the child types from some container type
+    such as `tuple[int, str]` and "pulls them out". For example,
+    `dict[Unpack[tuple[int, str]]` is equivalent to `dict[int, str]`.
+    In newer versions of Python, this is implemented using the `*` operator.
+
+    The type unpack operator can be applied to a `TypeVarTuple` instance:
+
+      Ts = TypeVarTuple('Ts')
+      tuple[Unpack[Ts]]  # Equivalent to tuple[*Ts]
+
+    Or to a parameterised `tuple`:
+
+      tuple[Unpack[tuple[int, str]]]  # Equivalent to tuple[*tuple[int, str]]
+    """
+
+    def __new__(cls, *args, **kwargs):
+        raise TypeError("Unpack should be used as Unpack[something] rather "
+                        "than Unpack(something)")
+
+    def __class_getitem__(cls, item):
+        if isinstance(item, types.GenericAlias) and item.__origin__ is tuple:
+            # tuple[...]
+            return StarredTuple(item)
+        elif isinstance(item, _GenericAlias) and item.__origin__ is tuple:
+            # typing.Tuple[...]
+            return StarredTuple(item)
+        elif isinstance(item, TypeVarTuple):
+            return item._unpacked
+        else:
+            raise TypeError(
```

Comment:
I'd prefer we don't raise an error here, so we don't close the door for other potential uses of `Unpack`. In particular, I'm thinking of making `Unpack[TypedDict type]` an allowed construct in the future. Another use case I can see is unpacking a subclass of tuple, such as a namedtuple type. 

In general, I want to avoid making the typing runtime too restrictive, so that users have the freedom to experiment with new uses of type annotations. We can disallow such unexpected uses in type checkers instead, where it's much easier to relax restrictions in the future.

Reply:
> In general, I want to avoid making the typing runtime too restrictive, so that users have the freedom to experiment with new uses of type annotations. We can disallow such unexpected uses in type checkers instead, where it's much easier to relax restrictions in the future.

Hmm, I see.

This seems like a strong argument to re-implementing `Unpack` using `_SpecialForm` - so that we actually know what to do if `Unpack` is called with something that isn't a `TypeVarTuple`, `tuple` or `Tuple`. I've gone ahead and done that (will comment in the other thread) and removed this error.